rct_image_tools/include/rct_image_tools/charuco_finder.h

```diff
@@ -0,0 +1,66 @@
+/**
+  * ChAruco gridboard detector, following the same pattern as ModifiedCircleGridObservationFinder.
+  * Author: John Berkebile
+  */
+
+#ifndef CHARUCO_FINDER_H
+#define CHARUCO_FINDER_H
+
+#include <boost/optional.hpp>
+#include <opencv2/core.hpp>
+#include <opencv2/aruco.hpp>
+#include <rct_optimizations/types.h>
+#include <map>
+#include <Eigen/Core>
+#include <iostream>
+
+namespace rct_image_tools
+{
+
+/**
+ * @brief This class finds 2D observations from images of a specified ChArUco gridboard target.
+ * The main advantage of this kind of target is that partial views still provide usable correspondences.
+ * Observations are returned as a map where the marker ID is the key and the image coordinates of the
+ * marker corners are the mapped value.
+ */
+class CharucoGridBoardObservationFinder
+{
+public:
+  CharucoGridBoardObservationFinder(const cv::Ptr<cv::aruco::CharucoBoard>& board);
+
+  /**
+   * @brief findObservations Detect marker corner coordinates in the provided image.
+   * @param image Input image, ideally containing an ArUco gridboard.
+   * @return Map matching marker ID numbers to a vector of marker corner coordinates. The vector will contain
+   * four corners defined in the same order as in the output of the function cv::aruco::DetectMarkers()
+   * (e.g. clockwise from the "origin" corner).
+   */
+  boost::optional<std::map<int, std::vector<Eigen::Vector2d>>> findObservations(const cv::Mat& image) const;
```

Comment:
I would remove the `boost::optional` from the return type and just return a map. If there are failures within this function, please throw an exception instead